appsmithorg/appsmith
# fix: 11468 select widget performance on a large data set.

## Description

- Improvements in initial render time of Select Widget on large data set.
- The original complaint was that in case of large data set, the widget takes up to 5 seconds to render.

Updates -
1. Biggest improvement gained by manually controlling the visibility of the menu list through a state variable. Prevents rendering of each item (total = 1000) on every parent re-render (even when the menu is hidden).
2. Cutting down unnecessary re-renders through memoization of child components and additional checks before calling setState.
3. Initial render is now down to ~100ms once the options are available from the api call.
4. Case - Opening the options menu for the first time. This was taking about 1.5s for a 1000 option list, mainly because of cascading updates and expensive style calculations in blueprint Text component. The only way to fix this was to create a custom MenuItem component. The rendering time is now down to ~700ms.

<img width="1680" alt="Screen Shot 2022-03-07 at 12 54 24 PM" src="https://user-images.githubusercontent.com/5424788/156992050-57b40e6f-c46e-44c8-bd5e-a5b6a95143c2.png">


Fixes #11468

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

1. Add a Select Widget to a page. 
5. Fetch options through an api call. Use the following api to get a 1000 item list - ```https://demo7762119.mockable.io/dropdown```.
6. Set default value to the first item [0].
7. Re-load the page and observe that the widget renders in under 100ms after the options are available from the api call. Use React profiler or performance dev tools to observe render times.

## Checklist:























## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>


    // Code coverage diff between base branch:release and head branch: fix-11468_dropdown_performance 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 55.69 **(-0.03)** | 37.07 **(-0.04)** | 35.29 **(0)** | 55.95 **(-0.02)**
 :red_circle: | app/client/src/utils/WorkerUtil.ts | 89.76 **(0)** | 70.59 **(-1.96)** | 100 **(0)** | 93.33 **(0)**
 :green_circle: | app/client/src/utils/helpers.tsx | 67.58 **(0.6)** | 35.86 **(1.5)** | 53.45 **(2.57)** | 65.48 **(0.68)**
 :red_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.71 **(-0.23)** | 40.83 **(-0.84)** | 36.21 **(0)** | 56.74 **(-0.25)**
 :sparkles: :new: | **app/client/src/widgets/SelectWidget/component/SelectButton.tsx** | **100** | **50** | **100** | **100**
 :green_circle: | app/client/src/widgets/SelectWidget/component/index.styled.tsx | 48.48 **(1.6)** | 25.81 **(2.48)** | 0 **(0)** | 51.61 **(1.61)**
 :red_circle: | app/client/src/widgets/SelectWidget/component/index.tsx | 14.52 **(-7.22)** | 0 **(0)** | 4.55 **(-1.01)** | 15.04 **(-6.14)**</details>
